homeassistant/components/unifiprotect/media_source.py

```diff
@@ -0,0 +1,862 @@
+        duration = end - start
+        title += f" {_format_duration(duration)}"
+        if event_type == EventType.RING.value:
+            event_text = "Ring Event"
+        elif event_type == EventType.MOTION.value:
+            event_text = "Motion Event"
+        elif event_type == EventType.SMART_DETECT.value:
+            if isinstance(event, Event):
+                smart_type = event.smart_detect_types[0]
+            else:
+                smart_type = SmartDetectObjectType(event["smartDetectTypes"][0])
+            event_text = f"Smart Detection - {smart_type.name.title()}"
+        title += f" {event_text}"
+
+        nvr = data.api.bootstrap.nvr
+        if thumbnail_only:
+            return BrowseMediaSource(
+                domain=DOMAIN,
+                identifier=f"{nvr.id}:eventthumb:{event_id}",
+                media_class=MEDIA_CLASS_IMAGE,
+                media_content_type="image/jpeg",
+                title=title,
+                can_play=True,
+                can_expand=False,
+                thumbnail=async_generate_thumbnail_url(
+                    event_id, nvr.id, THUMBNAIL_WIDTH, THUMBNAIL_HEIGHT
+                ),
+            )
+
+        return BrowseMediaSource(
+            domain=DOMAIN,
+            identifier=f"{nvr.id}:event:{event_id}",
+            media_class=MEDIA_CLASS_VIDEO,
+            media_content_type="video/mp4",
+            title=title,
+            can_play=True,
+            can_expand=False,
+            thumbnail=async_generate_thumbnail_url(
+                event_id, nvr.id, THUMBNAIL_WIDTH, THUMBNAIL_HEIGHT
+            ),
+        )
+
+    async def _build_events(
+        self,
+        data: ProtectData,
+        start: datetime,
+        end: datetime,
+        camera_id: str | None = None,
+        event_type: EventType | None = None,
+        reserve: bool = False,
+    ) -> list[BrowseMediaSource]:
+        """Build media source for a given range of time and event type."""
+
+        if event_type is None:
+            types = [
+                EventType.RING,
+                EventType.MOTION,
+                EventType.SMART_DETECT,
+            ]
+        else:
+            types = [event_type]
+
+        sources: list[BrowseMediaSource] = []
+        events = await data.api.get_events_raw(
+            start=start, end=end, types=types, limit=data.max_events
+        )
+        events = sorted(events, key=lambda e: cast(int, e["start"]), reverse=reserve)
+        for event in events:
+            # do not process ongoing events
+            if event.get("start") is None or event.get("end") is None:
+                continue
+
+            if camera_id is not None and event.get("camera") != camera_id:
+                continue
+
+            # smart detect events have a paired motion event
+            if (
+                event.get("type") == EventType.MOTION.value
+                and len(event.get("smartDetectEvents", [])) > 0
```

Comment:
We don't need to check length explicitly on containers if we just want to know that they are not empty. Containers will be truthy when they are not empty.

Reply:
#76870